### DISCUSSION TITLE
[API Request] Support timeOffset for audio when transcoding

### DISCUSSION THREAD

**Discussion by Tolriq** [opensubsonic/open-subsonic-api#21] [API requests]
### Type of change

API extension

### Proposal description

When transcoding, `estimateContentLength`  is not good enough to allow proper playback and seeking by most players.

To allow seeking by restarting the transcoding at the proper position it would be nice that the `timeOffset` parameter of `stream` was supported.

This is the case for all other providers supporting transcoding like Plex, Jellyfin, ...

### Backward compatibility impact

There's no backward impact, but it needs to be done as an extension so that clients knows that they can use it.

### Backward compatibility

- [X] No backward compatibility impact.

### API details

Just use the parameter as expected for video.

### Security impacts

N/A

### Potential issues

N/A

### Alternative solutions

I can't think of anything due to how on the fly transcoding works.

**Comment by spl0k** [comment]
Shouldn't this also be handled when there's no transcoding?
I don't know a thing about audio file formats so maybe this is not relevant but assuming they don't store some kind of timestamp->file offset mapping even without transcoding the actual file size can't be used for seeking on formats with a variable bitrate...

**Reply by Tolriq** [reply]
All the players can handle that by having the exact file size or for most format by counting the number of blocs or a special bloc at the end of the file.
During transcoding the filesize is unknown and the end of the file does not exist yet when playback starts.
And most important you can request to seek to a place that is not yet transcoded, while with a full file that new place is always already available.

Furthermore to support that it would require transcoding to actually be able to start at that point.

**Reply by spl0k** [reply]
Maybe I got you wrong but it seems to me you are assuming that if there's no transcoding the user already has the full file, which might not be the case. If you're streaming, again with no transcoding, on a poor connection you might just have a few samples available ahead of what you're playing, so you can't use that eventual special block at the end and you can also seek to a time position you haven't downloaded yet.

**Reply by Tolriq** [reply]
The server have the full file :)
So the client can request the proper range to get the data where it want and the headers do return the proper filesize since it's known.

**Reply by spl0k** [reply]
My point is if you restrict the `timeOffset` parameter to a transcoding context, the client doesn't have a way to request the proper range. Even with the file size available you can't accurately convert a time offset to a file offset when dealing with files with a variable bitrate.

**Reply by sentriz** [reply]
> All the players can handle that by having the exact file size

but with estimateContentLength, shouldn't the server respond with a valid Content-Length header? and actually send that number of bytes too (perhaps padded with zeros if the actual transcoded data ends up being smaller)

in that world, can't the client just send a HTTP Range request in order to seek?

**Reply by Tolriq** [reply]
With all due respect I code client apps playing audio since more than a decade.
Every app can currently already seek when playing from Subsonic without transcoding yes? This already works because of what I explained earlier. Else no one would be able to seek and this option would have been added since the start for audio too.

Since the server have the file the user can request the range they want inside the file via the header ranges (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range)
Since there's no transcoding the size and all the file is accessible from the client that can request what it want in the file.

Depending on the media type there's mechanism with that to be able to seek in files. Typically as explained earlier a lot of format write the information at the end of the file (moov for MP4 for example, for ogg reading the last block allows to know the number of granule and so generate the seek map) this end of file is accessible already since the file size is known so the client just request that bloc and can seek precisely (Something not true during transcoding that why the option is named `estimateContentLength` since it's know properly known). For some other format there's other ways.
But with proper filesize and range support nearly everything can already be seeked properly.
Most clients do request the start of the file, confirm the media type, read what they need to be able to load and seek (so usually also the last bytes of the file) then everything works.

Anyway you are free to offer `timeOffset` outside of transcoding but this brings nothing to clients and force servers to transcode a file to be able to serve from that starting point, this also means that the file duration will now be truncated and so that the client needs to handle this properly by knowing the duration and doing the proper UI handling for that. 

TL;DR; "I don't know a thing about audio file formats" probably is the cause of your concern. Just understand that clients can already seek with Subsonic when not transcoding so this probably means this is not needed when no transcode occurs.

**Reply by Tolriq** [reply]
> > All the players can handle that by having the exact file size
> 
> but with estimateContentLength, shouldn't the server respond with a valid Content-Length header? and actually send that number of bytes too (perhaps padded with zeros if the actual transcoded data ends up being smaller)
> 
> in that world, can't the client just send a HTTP Range request in order to seek?

No they can't because the content length is not known so is estimated. Transcoding is not finished when the file starts to play so there's no end of file to seek to anyway. Even if you used a fixed bitrate and a proper filesize. The client would ask for the last part of the file and the server would fail to serve it until the file have finished transcoding.

Serving file before transcoding is finished is important for ressource and speed you can't wait for a 2 hours mix to be transcoded before hearing the first seconds.

There's some specific format that are made to manage that, but even then to be able to seek at 1 hour mark it would still require that the file has transcoded past that point something that is not guaranteed at all specially on low power devices.

The `timeOffset` to restart playback at the proper point does address all the needs and that's why it's supported by all other providers who support transcoding.

**Reply by gravelld** [reply]
> The server have the full file :)

This may not be the case (ideally Astiga would only stream from backing storage service without downloading the full file eagerly), but it feels like there are workarounds.

**Reply by Tolriq** [reply]
> This may not be the case (ideally Astiga would only stream from backing storage service without downloading the full file eagerly), but it feels like there are workarounds.

I would expect that astiga act as a proxy in that case, so the filesize is the proper one and the http range requests are forwarded.

Having the full file does not mean on the local drive just that the file is complete (since no transcoding) and seekable (as seek in bytes not time position).

There might be a few very edge case where this is not the case. But to properly handle those we should first find a way to warn the clients about that case then decide how to handle this if there's the need to provide a workaround.

**Comment by Tolriq** [comment]
Implementation at https://github.com/opensubsonic/open-subsonic-api/pull/54

**Comment by dweymouth** [comment]
I have no problem with supporting the `timeOffset` property for music, but I think it should be possible as well for servers to treat HTTP range requests the same way as the timeOffset parameter. IE when transcoding, if the server gets a GET request for the file with a range header, if the start range is outside of the section of the file that is currently transcoded, the server can spin up a new transcode starting from the new position and serve that.

This means clients don't need to know whether they're being served a transcoded or non-transcoded file, they can just seek it and it will work

**Reply by dweymouth** [reply]
The reason to advocate for this is a lot of playback engines, like mpv, already handle seeks with range requests. Most servers already handle range requests properly for non-transcoded or fully-transcoded files.

**Reply by Tolriq** [reply]
Except that: 

1) When transcoding there's no range since we do not know the end :) The estimateContentLength produce wrong values that break playback at the end often.
2) Most renderers will try to seek to the end of the file to try to read possible metadata depending on the container format that would trigger erronous unwanted transcode restart that would break the first one. Or duplicate simultaneous transcode that would require the server to better handle them.

That's the whole point of this, it is not possible to properly manage transcoding seeking with http range. All other servers that support transcoding have a way to define a starting point to allow seeking or use HLS that works also differently. Http range is never the solution here.

**Reply by dweymouth** [reply]
Ah yeah I see why without an accurate knowledge of the content length seeking by byte offset can be way off. Since you may have already implemented timeOffset in Symfonium for other providers, can I ask about some implementation details? - how do you keep track of how much of the file in terms of playback time has been downloaded, so you can know if a seek is within the already-loaded portion and whether you need to execute a new HTTP request at all?

**Reply by Tolriq** [reply]
You can't know :)
Any seek in the future is a restart, only known valid range is between the starting point (so 0 or the seek point) and current position.

**Reply by dweymouth** [reply]
Why should that be the case though? If the server can transcode and client can download the file in faster than realtime, seeking forward but still within the already-downloaded portion shouldn't restart, right?

**Reply by dweymouth** [reply]
It seems if nothing else, the client could read from the data in two parallel tasks - one in realtime to send to the sound card, and another as fast as content is available just to keep track of how much is actually available to seek forward without issuing a new HTTP request?

I haven't tested with Wireshark or anything but I think mpv already does something like this automatically

**Reply by Tolriq** [reply]
But you can't know the status of the transcoding without an API, and some transcoder only transcode in real time when reading you can't make any assumptions that covers all the cases. 
And clients usually do not download the whole file they have buffers to pre load a part of the future playback don't keep 2 hours of FLAC data in memory specially on a phone.

**Reply by dweymouth** [reply]
Yeah, so if the forward-seek is within that buffer then you don't issue a new HTTP request, right? You only start a new transcode if the seek oversteps the buffer

**Reply by dweymouth** [reply]
I'm only asking all this because it seems like to use `timeOffset` I'll have to switch from mpv to using something custom and more low-level so just trying to get a sense of the implementation details of how to effectively use timeOffset

**Reply by Tolriq** [reply]
In theory yes, but if you seek too close of the end of the buffer you don't have enough buffer to actually play something or will play like 1 sec of data then pause until transcoder catch up.

This kind of optimisation will only bring you troubles for no benefits, the user transcode, seeking takes a little longer but works always the same without issues is the best user experience here.

You need to analyze buffer size, duration of playback for each buffer size, speed of filling the buffer and other parameter to eventually gain 0.5s sometimes on a seek.

You can try if you want but I would not recommend ;)

**Reply by Tolriq** [reply]
You don't have to change the renderer if you control the seek bar and position display.
But client support is out of scope here.

**Reply by Tolriq** [reply]
@dweymouth please do not close the issues there's the tags to navigate by status already and this just hide information by default to any newcomer.
